core/server/data/migrations/versions/1.23/1-update-koenig-beta-html.js

```diff
@@ -0,0 +1,49 @@
+const _ = require('lodash');
+const common = require('../../../../lib/common');
+const converters = require('../../../../lib/mobiledoc/converters');
+const models = require('../../../../models');
+
+const mobiledocIsCompatibleWithV1 = function mobiledocIsCompatibleWithV1(doc) {
+    if (doc
+        && doc.markups.length === 0
+        && doc.cards.length === 1
+        && doc.cards[0][0].match(/(?:card-)?markdown/)
+        && doc.sections.length === 1
+        && doc.sections[0].length === 2
+        && doc.sections[0][0] === 10
+        && doc.sections[0][1] === 0
+    ) {
+        return true;
+    }
+
+    return false;
+};
+
+module.exports.up = function regenerateKoenigBetaHTML(options) {
+    let postAllColumns = ['id', 'html', 'mobiledoc'];
+
+    let localOptions = _.merge({
+        context: {internal: true}
+    }, options);
+
+    common.logging.info('Migrating Koenig beta post\'s HTML to 2.0 format');
+
+    return models.Post.findAll(_.merge({columns: postAllColumns}, localOptions))
```

Comment:
Have you run a bigger database agains this migration script?
If not, i can do :) Just let me know. I am just curious how long it takes because of the extra line

> let html = converters.mobiledocConverter.render(mobiledoc, version);

Reply:
For my sqlite dev database which has ~90 Koenig beta posts in it:

`knex-migrator migrate --force  2.58s user 0.39s system 107% cpu 2.760 total`

It also threw quite a few errors though because I have posts with invalid mobiledoc in them.

I wonder how many people have large amounts of Koenig beta posts at the moment? The `html` conversion will only happen for those posts.

Reply:
> I wonder how many people have large amounts of Koenig beta posts at the moment? The html conversion will only happen for those posts.

Yeah agree. Maybe we shouldn't worry about the time.

> It also threw quite a few errors though because I have posts with invalid mobiledoc in them due to old test posts with card names that no longer exist.

What happened with them?
I've also realised yesterday that we might need a protection in the Ghost importer for invalid mobiledoc formats.

Reply:
> What happened with them?

With the posts that threw errors due to unknown mobiledoc cards? The `html` is still generated, it's just missing the content for any cards that didn't match up to our core cards.

Error output looks like this:

```
[2018-07-19 11:05:26] ERROR

NAME: InternalServerError
MESSAGE: Mobiledoc card 'card-hr' not found.

level:critical

InternalServerError: Mobiledoc card 'card-hr' not found.
    at new InternalServerError (/Users/kevinansfield/code/ghost/Ghost/node_modules/ghost-ignition/lib/errors/index.js:71:23)
    at Object.unknownCardHandler [as render] (/Users/kevinansfield/code/ghost/Ghost/core/server/lib/mobiledoc/converters/mobiledoc-converter.js:11:30)
    at Renderer.renderCardSection (/Users/kevinansfield/code/ghost/Ghost/node_modules/mobiledoc-dom-renderer/dist/commonjs/mobiledoc-dom-renderer/renderers/0-3.js:350:27)
    at Renderer.renderSection (/Users/kevinansfield/code/ghost/Ghost/node_modules/mobiledoc-dom-renderer/dist/commonjs/mobiledoc-dom-renderer/renderers/0-3.js:144:23)
    at /Users/kevinansfield/code/ghost/Ghost/node_modules/mobiledoc-dom-renderer/dist/commonjs/mobiledoc-dom-renderer/renderers/0-3.js:102:31
    at Array.forEach (<anonymous>)
    at Renderer.render (/Users/kevinansfield/code/ghost/Ghost/node_modules/mobiledoc-dom-renderer/dist/commonjs/mobiledoc-dom-renderer/renderers/0-3.js:101:21)
    at RendererFactory.render (/Users/kevinansfield/code/ghost/Ghost/node_modules/mobiledoc-dom-renderer/dist/commonjs/mobiledoc-dom-renderer/renderer-factory.js:108:71)
    at Object.render (/Users/kevinansfield/code/ghost/Ghost/core/server/lib/mobiledoc/converters/mobiledoc-converter.js:104:33)
    at Promise.map.concurrency (/Users/kevinansfield/code/ghost/Ghost/core/server/data/migrations/versions/1.25/1-update-koenig-beta-html.js:41:62)
    at tryCatcher (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/util.js:16:23)
    at MappingPromiseArray._promiseFulfilled (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/map.js:61:38)
    at MappingPromiseArray.PromiseArray._iterate (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/promise_array.js:114:31)
    at MappingPromiseArray.init (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/promise_array.js:78:10)
    at MappingPromiseArray._asyncInit (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/map.js:30:10)
    at Async._drainQueue (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/async.js:138:12)
    at Async._drainQueues (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/Users/kevinansfield/code/ghost/Ghost/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:794:20)
    at tryOnImmediate (timers.js:752:5)
    at processImmediate [as _immediateCallback] (timers.js:729:5)
```

> I've also realised yesterday that we might need a protection in the Ghost importer for invalid mobiledoc formats.

Possibly but I'm not sure how the mobiledoc format would be invalid unless it's been corrupted somehow, it's been inserted manually via the API, or it's a developer install that has posts around from the pre-1.0 alphas with a corrupted mobiledoc.

For unknown cards we currently throw errors but it's not really an end-the-world situation, maybe we should just log the problem instead?

Reply:
> For unknown cards we currently throw errors but it's not really an end-the-world situation, maybe we should just log the problem instead?

log sounds good if an unknown card is detected. I guess "unknown" means: i insert a random card name via the API manually?

I think it would be nice to have to add a validator, but lower prio. I agree it should only happen if you do something manually, but maybe nice for other external clients?
- you can't insert invalid mobiledoc structure via API
- you can't import invalid mobiledoc structure via importer 
- but we don't auto-fix

Reply:
> - you can't insert invalid mobiledoc structure via API

This should already error if the converter can't parse the mobiledoc. I was thinking more about people inserting stuff directly into the database, I don't think there's much we can do about that and it's definitely edge case.